### PR TITLE
add pipeline controller, take common controllers into a pipeline in order

### LINKED ISSUE #2449: applied chaos cannot be recovered if the CR is deleted before adding finalizers
## Bug Report

**What version of Kubernetes are you using?**

**What version of Chaos Mesh are you using?**

1.2.x and 2.0.x

**What did you do?**

Delete an XXXChaos instance.

**What did you expect to see?**

All injected faults are recovered.

**What did you see instead?**

CR is deleted, by faults are still injected.

**Possible Reasons**

In the 1.2.x case, if the chaos is being reconciled and is being applied, and users delete the chaos, the chaos will be deleted without recovering (because the finalizer hasn't been set). The way to fix this problem is to add a finalizer before applying the chaos.

In the 2.0.x case, this problem is not so obvious, as the chaos will not be applied multiple times (so that, only the very beginning and resuming from pause will run into this bug), but running the finalizer controller before the others will also make sure it doesn't happen.

### PR DESCRIPTION
**Pull request by Hexilee**
### What problem does this PR solve?

Close #2449

### What's changed and how it works?

Take common controllers into a pipeline in order.

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Chaos Dashboard`
- [x] Need to cheery-pick to release branches 2.0 and 2.1

### Checklist

Tests

- [x] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### CODE REVIEW

**Comment by STRRL** [comment]
What about explaining the cause of the issue in detail? :sob

I could only guess that executing these different reconcilers parallelly brings some unexpected behavior, from your code changes.

**Comment by YangKeao** [comment]
> What about explaining the cause of the issue in detail? sob
> 
> I could only guess that executing these different reconcilers parallelly brings some unexpected behavior, from your code changes.

Running controllers one by one will also resolve most chances of the conflict of one resource.

**Comment by Hexilee** [comment]
> What about explaining the cause of the issue in detail? 😭
> 
> I could only guess that executing these different reconcilers parallelly brings some unexpected behavior, from your code changes.

OK, I will supplement the details tomorrow.

**Comment by cwen0** [comment]
@Hexilee This issue has been fixed in #2509, Can we close this pr now?

**Comment by STRRL** [comment]
please resolve the conflicts @Hexilee

**Comment by cwen0** [comment]
@Hexilee CI was broken! Can you fix it?

**Comment by Hexilee** [comment]
> @Hexilee CI was broken! Can you fix it?

Fixed

File: controllers/common/controller.go [hunk 1]
```diff
@@ -27,18 +27,14 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (YangKeao) [inline controllers/common/controller.go:27]: How about move this controller into a standalone directory? (as one of the subdirectory of `common`).

Though it was named `common`, it was a standalone controller "records" like anyone else ("condition", "desiredphase"...)

Comment (YangKeao) [reply controllers/common/controller.go:27]: And the `README.md` should also be moved.

File: controllers/action/delegate.go [hunk 2]
```diff
@@ -23,15 +23,12 @@ import (
...
 	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
-	"github.com/chaos-mesh/chaos-mesh/controllers/common"
 )
 
 type Delegate struct {
 	impl interface{}
 }
 
-var _ common.ChaosImpl = (*Delegate)(nil)
-
```
Comment (STRRL) [inline controllers/action/delegate.go:34]: please add `var _ impltypes.ChaosImpl = (*Delegate)(nil)`

**Review by STRRL — COMMENTED** [review]
rest LGTM

File: controllers/common/pipeline/pipeline.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+			return ret, nil
+		}
+
+		if ret.RequeueAfter != 0 {
+			if result.RequeueAfter == 0 {
+				result.RequeueAfter = ret.RequeueAfter
+			} else {
+				result.RequeueAfter = minDuration(result.RequeueAfter, ret.RequeueAfter)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply controllers/common/pipeline/pipeline.go:92]: Yes, it's expected. As `Requeue` means re-queue as soon as possible, we should short-circuit the pipeline.

File: controllers/types/types.go [hunk 4]
```diff
@@ -24,7 +24,7 @@ import (
 type Controller string
 
 type Object struct {
-	Object v1alpha1.InnerObject
+	Object v1alpha1.InnerObjectWithSelector
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply controllers/types/types.go]: Fixed

File: controllers/action/delegate.go (hunk 2 above)
Comment (Hexilee) [reply controllers/action/delegate.go]: Fixed

File: controllers/common/controller.go (hunk 1 above)
Comment (Hexilee) [reply controllers/common/controller.go:27]: Fixed

File: controllers/common/pipeline/pipeline.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (STRRL) [reply controllers/common/pipeline/pipeline.go:92]: How about also letting `RequeueAfter` short-circuit the pipeline? Keep them the same. ❤️ 

The different behavior adds more logic on `Requeue` and `RequeueAfter`, but we still use the concepts provided by controller-time, which would confuse users, and might bring potential issues.

[the remark this answers is not in the document]
Comment (Hexilee) [reply controllers/common/pipeline/pipeline.go:92]: The `desiredphase` controller, which controls duration and excess, will always return `RequeueAfter` or `Requeue`. If we let `RequeueAfter` short-circuit the pipeline, the controllers behand `desiredphase` will never be executed.

[the remark this answers is not in the document]
Comment (STRRL) [reply controllers/common/pipeline/pipeline.go]: > 

Got that. How about writing the reason as comments at line 80?

[the remark this answers is not in the document]
Comment (STRRL) [reply controllers/common/pipeline/pipeline.go]: (Or could we split `desiredphase` out of `AllSteps`?

[the remark this answers is not in the document]
Comment (Hexilee) [reply controllers/common/pipeline/pipeline.go]: I will add comments. (I don't think `desiredphase` is special enough to be split out)

[the remark this answers is not in the document]
Comment (Hexilee) [reply controllers/common/pipeline/pipeline.go]: Some comments and a README.md are added.

**Review by STRRL — APPROVED** [review]
LGTM!

**Review by YangKeao — APPROVED** [review]
LGTM

**Comment by STRRL** [comment]
Hi, @Hexilee PTAL, please make manually cherry-picks on 2.1 and 2.0.

**Comment by Hexilee** [comment]
> Hi, @Hexilee PTAL, please make manually cherry-picks on 2.1 and 2.0.

WIP

**Pull request merged**